addon/serializers/rest.js

```diff
@@ -200,8 +200,8 @@ var RESTSerializer = JSONSerializer.extend({
     const primaryHasTypeAttribute = modelHasAttributeOrRelationshipNamedType(primaryModelClass);
     // Support polymorphic records in async relationships
     if (!primaryHasTypeAttribute && hash.type && store._hasModelFor(this.modelNameFromPayloadKey(hash.type))) {
-      serializer = store.serializerFor(hash.type);
-      modelClass = store.modelFor(hash.type);
+      serializer = store.serializerFor(this.modelNameFromPayloadKey(hash.type));
+      modelClass = store.modelFor(this.modelNameFromPayloadKey(hash.type));
```

Comment:
hmm yes, I see the confusion. The API that led me to find this bug is a weird pseudo-JSON API that was settled during the RC phase, it uses `type`. I was confused while debugging by the role that `JSONSerializer` inheritance played, but you probably know better than me whether it’s feasible to change the method name.